File: fsharp/house/House.fs
module House

let pattern = 
    [
        ("house","lay in")
        ("malt", "ate");
        ("rat", "killed");
        ("cat", "worried");
        ("dog","tossed");
        ("cow with the crumpled horn","milked");
        ("maiden all forlorn","kissed");
        ("man all tattered and torn","married");
        ("priest all shaven and shorn","woke");
        ("rooster that crowed in the morn","kept");
        ("farmer sowing his corn","belonged to");
        ("horse and the hound and the horn","")
    ]

let recite startVerse endVerse: string list =
    let lines x = 
        seq { 
            for i in [x-1..-1..0] do 
                yield 
                    match i with
                    | _ when i = 0 && x = 1 -> "This is the house that Jack built."
                    | _ when i = 0 -> "that lay in the house that Jack built."
                    | _ when i = x-1 -> sprintf "This is the %s" (fst pattern.[i])
                    | _ -> sprintf "that %s the %s" (snd pattern.[i]) (fst pattern.[i])
            if x <> endVerse then yield ""
        }
    [startVerse..endVerse] |> Seq.collect lines |> Seq.toList

File: csharp/house/House.cs
using System;
using System.Linq;

public static class House
{
    private static readonly (string thing, string did)[] verses = new[]
    {
        ("",""),
        ("house that Jack built.", ""),
        ("malt", "lay in"), 
        ("rat", "ate"), 
        ("cat", "killed"), 
        ("dog", "worried"),
        ("cow with the crumpled horn", "tossed"), 
        ("maiden all forlorn", "milked"),
        ("man all tattered and torn", "kissed"), 
        ("priest all shaven and shorn", "married"),
        ("rooster that crowed in the morn", "woke"), 
        ("farmer sowing his corn", "kept"),
        ("horse and the hound and the horn", "belonged to")
    };

    public static string Recite(int verseNumber)
    {
        if(verseNumber < 1 || verseNumber > 12)
            throw new ArgumentException(nameof(verseNumber));

        var result = "";
        for(var i = verseNumber; i > 0; i--)
            result += i == verseNumber
                ? $"This is the {verses[i].thing}"
                : $" that {verses[i + 1].did} the {verses[i].thing}";
        return result;
    }

    public static string Recite(int startVerse, int endVerse) =>
        string.Join("\n", 
            Enumerable
                .Range(startVerse, endVerse - startVerse + 1)
                .Select(i => Recite(i)));
}